docs/ansi_seq.man
.\" USMID %W%
.\" RCSID @(#)$Id: ansi_seq.man,v 1.1 1998/10/07 16:19:56 rk Exp $
.\" LIBDIR
.TH "ANSI_SEQ" "3rko" "12 May 1995"
.SH NAME
ansi_seq \- C functions that return an ANSI sequence to manipulate
the screen character attributes and colors.
.SH SYNOPSIS

 #include "librko.h"

 #define ATTRIB_NUM	6
 typedef enum {
   NONE,   BOLD,   UNDERSCORE,   BLINK,   REVERSE,   CONCEALED}
 ansi_attributes;

 #define COLOR_NUM	8
 typedef enum {
   BLACK,   RED,   GREEN,   YELLOW,   BLUE,   MAGENTA,   CYAN,   WHITE}
 ansi_colors;

 extern const char *ANSI_Attributes[ATTRIB_NUM];
 extern const char *ANSI_Colors[COLOR_NUM];

 #define ANSISEQ_LEN	16
 #define ANSISEQ_NUM	32

 int set_ansi_tty(int ans);
 const char *ansi_seq(ansi_attributes a, ansi_colors b, ansi_colors f);
 const char *ansi_color(ansi_colors b, ansi_colors f);
 const char *ansi_fgcolor(ansi_colors f);
 const char *ansi_bgcolor(ansi_colors b);
 const char *ansi_attribute(ansi_attributes a);

.SH DESCRIPTION
ANSI sequences are of the form ``^[[xx;xx;xxm'', where ``^['' is
an escape character and ``xx'' is the ANSI code to set either the
foreground color, background color, or the character attributes (e.g.
bold, reverse,
.IR .etc ).
If, however, the standard output is not tied to a tty device then the
following ANSI sequence generating functions will be disabled and will
return a char * pointer to a '\0', which will signal an empty string
in most output or string handling routines.  In other words, if the
stdout is redirected or piped to another process then the ANSI sequences
are disabled.

.I set_ansi_tty
will enable/disable the ANSI sequences given whether the input is
TRUE or FALSE respectively.  (TRUE is non-zero, FALSE is zero.)
The function returns the previous state.

.I ansi_seq
sets attribute, background color, and the foreground color in one sequence.
Note that it returns a pointer to an internal storage area which gets cycled
through with each call.  There are ANSISEQ_NUM internal buffer areas.
This allows multiple calls to ansi_seq() within the same printf()
statement.

.I ansi_color
sets only the background color and the foreground color.  It behaves in
the same manner as
.IR ansi_seq .

.I ansi_fgcolor
sets only the character foreground color.  It returns a pointer to an internal
storage area, which is unique to each foreground color sequence.

.I ansi_bgcolor
sets only the character background color.  It returns a pointer to an internal
storage area, which is unique to each background color sequence.

.I ansi_attribute
sets only the character attributes.  It returns a pointer to an internal
storage area, which is unique to each attribute sequence.

.I ANSI_Colors
and
.I ANSI_Attributes
are char arrays that describe the corresponding ANSI color or attribute.

.SH EXAMPLE
    int i
    ...
    for (i = 0; i < COLOR_NUM; ++i) {
        (void) printf("%s%-20s%s %d\n",
            ansi_seq(NONE, WHITE, MAGENTA),
            ANSI_Colors[i],
            ansi_seq(NONE, BLACK, WHITE),
            i);
    }

This code fragment would print out the color names with MAGENTA characters
on a WHITE background, followed by the color number with ordinary
white numbers on a black background.  Since the buffer pointers are
cycled in each call the second call to ansi_seq() returns a pointer to
a char * different from the first call.

.SH SEE ALSO
printf(3), strcpy(3)

.SH BUGS
There are only a finite number of internal buffers given by ANSISEQ_NUM.
It is therefore possible to have ``too many'' calls to ansi_seq() such
that a subsequent call overwrites a buffer area being pointed to by
a char * ptr, thus changing the ANSI sequence attributes and colors.
This is not a problem unless the programmer is expecting a specific
sequence to exist at a later time.  This will result in surprising
and unexpected attributes and colors.
Therefore, as in the above example, it's advisable to keep the calls
to ansi_seq() ``local'' to the printf() statement, or to strcpy() the
sequence to a local area in the program for persistent sequences.

.SH AUTHOR
R.K.Owen,Ph.D.

.KEY WORDS


NEWS

librko-1.0.0	2002/10/21
* first release with a working configure, using automake/autoconf etc.
* added spawn3



ansi_seq.c
static const char RCSID[]="@(#)$Id: ansi_seq.c,v 1.3 1999/03/25 15:08:31 rk Exp $";
static char AUTHOR[]="@(#)ansi_seq 1.1 12/21/95 R.K.Owen,PhD";

/* ansi_seq - creates a char * string that is the ANSI color and
 *	attribute sequence.
 * An ANSI sequence consists of one or more of the following numeric codes:
 * Attribute codes: 
 * "00"=none "01"=bold "04"=underscore "05"=blink "07"=reverse "08"=concealed
 * Text color codes:
 * 30=black 31=red 32=green 33=yellow 34=blue 35=magenta 36=cyan 37=white
 * Background color codes:
 * 40=black 41=red 42=green 43=yellow 44=blue 45=magenta 46=cyan 47=white
 *
 * an ANSI sequence is initiated with "^[[xx;xx;xxm"  where xx are the
 * values specified above and ^[ refers to an <esc>
 *
 * by R.K. Owen, Ph.D.  12/21/95
 */
/*
 *********************************************************************
 *
 *     This software is copyrighted by R.K.Owen,Ph.D. 1997
 *
 * The author, R.K.Owen, of this software is not liable for any
 * problems WHATSOEVER which may result from use  or  abuse  of
 * this software. The author, R.K.Owen, grants unlimited rights
 * to anyone who uses and modifies this  software  for  private
 * non-commercial  use  as  long  as  this copyright and rights
 * notice remains in this software and is made available to all
 * recipients of this software.
 *
 * last known email: rk@owen.sj.ca.us
 *                   rkowen@ckns.net
 *
 *********************************************************************
 */

#include <stdio.h>
#include <unistd.h>
#include "librko.h"

const char *ANSI_Attributes[ATTRIB_NUM] = {
  "none", "bold", "underscore", "blink", "reverse", "concealed"};
const char *ANSI_Colors[COLOR_NUM] = {
  "black", "red", "green", "yellow", "blue", "magenta", "cyan", "white"};

static char *attribnum[ATTRIB_NUM] = {
  "00",   "01",   "04",         "05",    "07",      "08"};
static char *fgcolornum[COLOR_NUM] = {
  "30",    "31",   "32",   "33",     "34",   "35",      "36",   "37"};
static char *bgcolornum[COLOR_NUM] = {
  "40",    "41",   "42",   "43",     "44",   "45",      "46",   "47"};

static int IS_TTY_OUT;

static void is_ansi_tty(void) {
#ifndef NO_ISATTY
	if (isatty(STDOUT_FILENO)) {
		IS_TTY_OUT = 1;
	} else {
		IS_TTY_OUT = -1;
	}
#else
	IS_TTY_OUT = 1;		/* always set for ANSI sequences */
#endif
}

int set_ansi_tty(int ans) {
	int t;
	if (!IS_TTY_OUT) is_ansi_tty();
	t = (IS_TTY_OUT + 1)/2;	/* saves it as 0 or 1 */
	if (ans)	IS_TTY_OUT = 1;
	else		IS_TTY_OUT = -1;
	return t;			/* returns the old setting */
}

const char *ansi_seq(ansi_attributes a, ansi_colors b, ansi_colors f) {
	static char buffer[ANSISEQ_NUM][ANSISEQ_LEN];
	static int current;	/* initialized to 0 */

	if (!IS_TTY_OUT) is_ansi_tty();
	if (IS_TTY_OUT == 1) {
		current++;
		current %= ANSISEQ_NUM;
		(void) sprintf(buffer[current], "\x1B[%s;%s;%sm",
			attribnum[a], bgcolornum[b], fgcolornum[f]);
		return (const char *) buffer[current];
	} else {
		*buffer[current] = '\0';
		return (const char *) buffer[current];
	}
}

const char *ansi_color(ansi_colors b, ansi_colors f) {
	static char buffer[ANSISEQ_NUM][ANSISEQ_LEN];
	static int current;	/* initialized to 0 */

	if (!IS_TTY_OUT) is_ansi_tty();
	if (IS_TTY_OUT == 1) {
		current++;
		current %= ANSISEQ_NUM;
		(void) sprintf(buffer[current], "\x1B[%s;%sm",
			bgcolornum[b], fgcolornum[f]);
		return (const char *) buffer[current];
	} else {
		*buffer[current] = '\0';
		return (const char *) buffer[current];
	}
}

const char *ansi_fgcolor(ansi_colors f) {
	static const char *fgcolorseq[COLOR_NUM] = {
	"\x1B[30m","\x1B[31m","\x1B[32m","\x1B[33m",
	"\x1B[34m","\x1B[35m","\x1B[36m","\x1B[37m"};
	static char buffer = '\0';

	if (!IS_TTY_OUT) is_ansi_tty();
	if (IS_TTY_OUT == 1) {
		return fgcolorseq[f];
	} else {
		return (const char *) &buffer;
	}
}

const char *ansi_bgcolor(ansi_colors b) {
	static const char *bgcolorseq[COLOR_NUM] = {
	"\x1B[40m","\x1B[41m","\x1B[42m","\x1B[43m",
	"\x1B[44m","\x1B[45m","\x1B[46m","\x1B[47m"};
	static char buffer = '\0';

	if (!IS_TTY_OUT) is_ansi_tty();
	if (IS_TTY_OUT == 1) {
		return bgcolorseq[b];
	} else {
		return (const char *) &buffer;
	}
}

const char *ansi_attribute(ansi_attributes a) {
	static const char *attribseq[ATTRIB_NUM] = {
	"\x1B[00m","\x1B[01m","\x1B[04m","\x1B[05m","\x1B[07m","\x1B[08m"};
	static char buffer = '\0';

	if (!IS_TTY_OUT) is_ansi_tty();
	if (IS_TTY_OUT == 1) {
		return attribseq[a];
	} else {
		return (const char *) &buffer;
	}
}
